docs/esp32: UDP MULTICAST client/server example.

IhorNehrutsa opened the pull request:
Use UDP MULTICAST connections to communicate between different networks.
For example between ESP32 with 172.16.55.55 IP address from B class network and 
ESP32 with 192.168.44.44 IP address from C class network.

Run `udp_multicast_client.py` on one device. Client send GET and receive ACK from the server.
Output is:
```
GET to   ('224.0.0.111', 5555)   sent     "client_ip:172.16.55.55  mac:b8-d6-1a-5a-c2-e4"
ACK from ('192.168.44.44', 5555) received "server_ip:192.168.44.44 mac:24-6f-28-7a-f3-9c"
GET to   ('224.0.0.111', 5555)   sent     "client_ip:172.16.55.55  mac:b8-d6-1a-5a-c2-e4"
ACK from ('192.168.44.44', 5555) received "server_ip:192.168.44.44 mac:24-6f-28-7a-f3-9c"
...
```

Run `udp_multicast_server.py` on another device. Server receive GET and send ACK to the client.
Output is:
```
GET from ('172.16.55.55', 5555) received "client_ip:172.16.55.55  mac:b8-d6-1a-5a-c2-e4"
ACK to   ('224.0.0.111', 5555)  sent     "server_ip:192.168.44.44 mac:24-6f-28-7a-f3-9c"
GET from ('172.16.55.55', 5555) received "client_ip:172.16.55.55  mac:b8-d6-1a-5a-c2-e4"
ACK to   ('224.0.0.111', 5555)  sent     "server_ip:192.168.44.44 mac:24-6f-28-7a-f3-9c"
...
```

IhorNehrutsa commented:
@DvdGiessen I've seen a few of your esp32/network PR`s. Maybe this will interest you.

DvdGiessen commented on docs/esp32/tutorial/udp_multicast.rst:
```diff
@@ -0,0 +1,141 @@
+    sockaddr = (IP, PORT)
+
+    wlan = network.WLAN(network.STA_IF)
+    ip = wlan.ifconfig()[0]
```
Having variable names only differ in case (`IP`  and `ip`) might not be the most readable for an example.

IhorNehrutsa replied:
Renamed "IP" to "MULTICAST_IP" and
"ip" to "client_ip"/"server_ip"

DvdGiessen commented on docs/esp32/tutorial/udp_multicast.rst:
```diff
@@ -0,0 +1,141 @@
+        try:
+            if time.time() - t >= 3:
+                t = time.time()
+                str_to_send = f'ip:{ip}\t mac:{wlan.config("mac")}'
```
You can format the MAC address for readability: `':'.join(map('{:02x}'.format, mac))`

IhorNehrutsa replied:
Done

DvdGiessen reviewed (commented):
I haven't (yet) had a need for doing anything multicast on MicroPython. Added two small comments on the example code; note I only glanced over it, didn't try it.

Is there anything MicroPython-specific about this (other than connecting the WiFi)? It doesn't look specific to the ESP32; this looks like it should work on any port (or even CPython) when these socket options are available.

IhorNehrutsa commented:
@DvdGiessen Thank you for the review.

dpgeorge commented:
> Is there anything MicroPython-specific about this (other than connecting the WiFi)? It doesn't look specific to the ESP32; this looks like it should work on any port (or even CPython) when these socket options are available.

Indeed, this kind of example/tutorial should be placed somewhere more generic.

Since the majority of this PR is code examples I suggest it be placed in `examples/network/`, as two separate files.

Commit:
examples/network: Add UDP multicast client and server examples.

examples/network: Add UDP multicast client and server examples.

Signed-off-by: IhorNehrutsa <Ihor.Nehrutsa@gmail.com>

IhorNehrutsa commented:
@dpgeorge Please squash 4 commits `examples/network: Add UDP multicast client and server examples.`  to one commit
and squash the other 6 commits to a second commit. Thanks

IhorNehrutsa commented:
@dpgeorge 
udp_multicast_*.py are placed to the examples/network/

IhorNehrutsa committed:
docs/esp32: UDP multicast client/server example.

Signed-off-by: IhorNehrutsa <Ihor.Nehrutsa@gmail.com>

Josverl commented on examples/network/udp_multicast_server.py:
```diff
@@ -0,0 +1,66 @@
+# udp_multicast_server.py::
+
+import network, socket, errno, struct, time
```
Feel free to ignore as this is really a nitpick: 
`time` is not used in the server example.

Josverl commented on docs/esp32/tutorial/udp_multicast.rst:
```diff
@@ -0,0 +1,27 @@
+.. _esp32_udp_multicast:
```
I think the better place for this is in a common location - rather than just for one port.
Currently the doc tree only has one common location , the reference section in `docs/reference` 
I suggest to move it there (, or possibly create a "common / shared" folder to reduce duplication - in a follow up PR) 

It would still make sense to link from the esp32 QS , and other networked QS pages to there 
(the same code tested just fine on a RPI_PICO_W)

Josverl reviewed (changes requested):
Tested the code on ESP32-C6 and RIPI_PICO_W- both worked fine 
I suggest the docs should be accessible / linked from other networked boards as well

IhorNehrutsa committed:
udp_multicast_server: Remove import time module.

Signed-off-by: IhorNehrutsa <Ihor.Nehrutsa@gmail.com>

IhorNehrutsa committed:
udp_multicast: Rename link.

Signed-off-by: IhorNehrutsa <Ihor.Nehrutsa@gmail.com>

IhorNehrutsa committed:
udp_multicast: Change location.

Move udp_multicast.rst to the docs/reference directory.
Register it in the index.rst.
Add link to the Rasberry Pi rp2/quickref.rst

Signed-off-by: IhorNehrutsa <Ihor.Nehrutsa@gmail.com>

IhorNehrutsa commented:
Thanks, @Josverl, for the review and tests.
Now rp2 `Networking` menu looks:
<img width="313" height="358" alt="image" src="https://github.com/user-attachments/assets/955a45c5-00f9-46a5-a7fd-e8cea263a867" />
esp32 `Networking` menu:
<img width="308" height="336" alt="image" src="https://github.com/user-attachments/assets/52215e8f-7d3c-4e82-ba2f-c036dc8b5bcc" />
